src/features/terminal/terminalManager.ts

```diff
@@ -93,37 +105,158 @@ export class TerminalManagerImpl implements TerminalManager {
+                    }
+                }),
+            );
+
+            if (shellsToSetup.length === 0) {
+                traceVerbose(`No shell profiles to setup for ${Array.from(shellTypes).join(', ')}`);
+                return;
+            }
+
+            if (!this.hasFocus) {
+                traceVerbose('Window does not have focus, skipping shell profile setup');
+                return;
+            }
+
+            setImmediate(async () => {
+                // Avoid blocking this setup on user interaction.
+                await handleSettingUpShellProfile(shellsToSetup, (p, v) => this.shellSetup.set(p.shellType, v));
+            });
+        }
+    }
+
+    private getShellActivationType(shellType: string): AutoActivationType | undefined {
+        let isSetup = this.shellSetup.get(shellType);
+        if (isSetup === true) {
+            traceVerbose(`Shell profile for ${shellType} is already setup.`);
+            return 'shellStartup';
+        } else if (isSetup === false) {
+            traceVerbose(`Shell profile for ${shellType} is not set up, using command fallback.`);
+            return 'command';
+        }
+    }
+
+    private async getEffectiveActivationType(shellType: string): Promise<AutoActivationType> {
+        const providers = this.startupScriptProviders.filter((p) => p.shellType === shellType);
+        if (providers.length > 0) {
+            traceVerbose(`Shell startup is supported for ${shellType}, using shell startup activation`);
+            let isSetup = this.getShellActivationType(shellType);
+            if (isSetup !== undefined) {
+                return isSetup;
+            }
+
+            await this.handleSetupCheck(shellType);
+
+            // Check again after the setup check.
+            return this.getShellActivationType(shellType) ?? 'command';
+        }
+        traceInfo(`Shell startup not supported for ${shellType}, using command activation as fallback`);
+        return 'command';
+    }
+
     private async autoActivateOnTerminalOpen(terminal: Terminal, environment: PythonEnvironment): Promise<void> {
-        const config = getConfiguration('python');
-        if (!config.get<boolean>('terminal.activateEnvironment', false)) {
-            return;
+        let actType = getAutoActivationType();
+        const shellType = identifyTerminalShell(terminal);
+        if (actType === 'shellStartup') {
+            actType = await this.getEffectiveActivationType(shellType);
         }
 
-        if (isActivatableEnvironment(environment)) {
-            await withProgress(
-                {
-                    location: ProgressLocation.Window,
-                    title: `Activating environment: ${environment.environmentPath.fsPath}`,
-                },
-                async () => {
-                    await waitForShellIntegration(terminal);
-                    await this.activate(terminal, environment);
-                },
+        if (actType === 'command') {
```

Comment:
might be worth having constants for all the activation types instead of the strings

Reply:
Great suggestion, will add in a flow up PR.